### PR TITLE
ARM full build pipeline

### LINKED ISSUE #12910: [DevOps] Create ARM build workflow

### PR DESCRIPTION
**Pull request by tmdzk**
ARM full build pipeline

## Summary

ARM full build pipeline

## Test Plan

Arm build pipeline to pass 

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: .github/workflows/build_all.yml [hunk 1]
```diff
@@ -893,6 +893,88 @@ jobs:
           SLACK_ICON_EMOJI: ":heavy_check_mark:"
           SLACK_COLOR: "#00FF00"
           SLACK_FOOTER: ' '
+  build-deploy-arm:
```
Comment (crasu) [inline .github/workflows/build_all.yml:896]: This still seems to fail with:
"Decoding ARM AGW artifacts JSON has failed:  "

The state on magma ci is not showing up as well.

**Comment by crasu** [comment]
@tmdzk 
This is publishing the new arm agw debs to this location (correct?):
https://artifactory.magmacore.org/artifactory/docker-test/agw_gateway_c_arm/
Is this a good run to verify the build is working?
https://github.com/magma/magma/runs/8277560165?check_suite_focus=true

File: lte/gateway/deploy/arm_build.yml [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+          assign_public_ip: yes
+        tags:
+          Environment: CI
+      register: ec2
+
+    - name: Sleep for 10 seconds if AWS didn't return the public ip
+      ansible.builtin.wait_for:
+        timeout: 10
+      delegate_to: localhost
+      when: ec2.instances[0].public_ip_address is undefined
```
Comment (Neudrino) [inline lte/gateway/deploy/arm_build.yml:62]: A static wait seems rather crude. Is there no better way to ensure a public IP has been assigned?

File: lte/gateway/deploy/arm_build.yml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        volumes:
+          - device_name: /dev/sda1
+            ebs:
+              volume_size: 50
+              delete_on_termination: true
+        network:
+          assign_public_ip: yes
+        tags:
+          Environment: CI
+      register: ec2
```
Comment (Neudrino) [inline lte/gateway/deploy/arm_build.yml]: This seems to be a rather large code duplication to https://github.com/magma/magma/pull/13889/files#diff-b64d719b0a8f19beec9889b3ada7c9a5bc5136edfcceb0f3e5bc4441b498db47R32-R52.
Why would that be necessary?

File: lte/gateway/deploy/arm_build.yml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      when: ec2.instances[0].public_ip_address is undefined
```
Comment (Neudrino) [inline lte/gateway/deploy/arm_build.yml:96]: What for an object is `ec2` exactly referring to?
Is this a Ansible internal variable?
Or is this a AWS resource (from some Ansible plugin)?

If the latter one is correct: How do you ensure that the first EC2 instance is the correct one? There could be other EC2 instances already in that account. Would that work in this case too?

File: lte/gateway/deploy/arm_build.yml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        groups: docker
+        append: yes
+      become: true
+
+    - name: Installing latest docker-compose
+      become: true
+      ansible.builtin.shell: |
+        COMPOSE_VERSION=$(curl -s "https://api.github.com/repos/docker/compose/tags" | jq -r '.[0].name')
+        curl -L "https://github.com/docker/compose/releases/download/${COMPOSE_VERSION}/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
+        chmod +x /usr/local/bin/docker-compose
```
Comment (Neudrino) [inline lte/gateway/deploy/arm_build.yml:158]: Why would this cumbersome manual installation be necessary, when we already installed docker-compse via [apt above](https://github.com/magma/magma/pull/13889/files#diff-b64d719b0a8f19beec9889b3ada7c9a5bc5136edfcceb0f3e5bc4441b498db47R122).

File: .github/workflows/build_all.yml [hunk 6]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+    runs-on: ubuntu-latest
+    env:
+      MAGMA_ROOT: "${{ github.workspace }}"
+      MAGMA_VERSION: "1.8.0"
+      MAGMA_AGW_CONTAINER_REGISTRY: "agw-test.artifactory.magmacore.org"
+    steps:
+      - uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # pin@v2
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:903]: Can we use the latest version 3, please.

File: .github/workflows/build_all.yml [hunk 7]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+      - name: Run apt
+        run: sudo apt-get update && sudo apt -y upgrade
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:905]: For which part of the pipeline would a complete upgrade of the runner be required?

File: .github/workflows/build_all.yml [hunk 8]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+      - uses: actions/setup-python@7f80679172b057fc5e90d70d197929d454754a5a # pin@v2
+        with:
+          python-version: '3.8.10'
+      - name: Setup SSH Keys and known_hosts
+        env:
+          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
+        run: |
+          ssh-agent -a $SSH_AUTH_SOCK > /dev/null
+          ssh-add - <<< "${{ secrets.LF_AWS_KEYPAIR }}"
+          mkdir ~/.ssh
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:919]: I would always use the `-p` option in scripts as best practice, even though this is not strictly necessary in this particular case.

File: .github/workflows/build_all.yml [hunk 9]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+          default: 3.8.10
+      - uses: actions/setup-python@7f80679172b057fc5e90d70d197929d454754a5a # pin@v2
+        with:
+          python-version: '3.8.10'
+      - name: Setup SSH Keys and known_hosts
+        env:
+          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
+        run: |
+          ssh-agent -a $SSH_AUTH_SOCK > /dev/null
+          ssh-add - <<< "${{ secrets.LF_AWS_KEYPAIR }}"
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:918]: This seems to be quite cumbersome.
From the rest of the implementation I get, that there is no ssh configured on the runners whatsoever. In that case it is much easier and (and imho) cleaner if we first create the key file and add it with `ssh-add` afterwards.
We could even align with the default name `id_rsa` to ease things up even more.

File: .github/workflows/build_all.yml [hunk 10]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+          mkdir ~/.ssh
+          touch ~/.ssh/gha_keypair.pem
+          echo "${{ secrets.LF_AWS_KEYPAIR }}" > ~/.ssh/gha_keypair.pem
+          sudo chmod 400 ~/.ssh/gha_keypair.pem
+      - name: Install Dependencies
+        run: |
+          pip install ansible awscli boto3
+          sudo apt-get update
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:930]: Why is this needed here?
It is also a duplication to [above](https://github.com/magma/magma/pull/13889/files#diff-71a71e993dac5ebcaa1ad57402f04ee226b14121a8052e4f994706670fd65a1aR905).

File: .github/workflows/build_all.yml [hunk 11]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+      - name: Get the right tagging
+        if: github.event_name == 'push'
+        run: |
+          echo TAG="${MAGMA_VERSION}" >> $GITHUB_ENV
+          if [ "$GITHUB_REF" = "refs/heads/master" ]; then
+            echo TAG="${GITHUB_SHA:0:8}" >> $GITHUB_ENV
+          fi
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:937]: An traditional `if ... else` statement would seem cleaner to me here, in contrast to overwriting the variable.

File: .github/workflows/build_all.yml [hunk 12]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+      - name: Ansible config setup
+        run: |
+          echo "[defaults]" >> ~/.ansible.cfg
+          echo "host_key_checking = False" >> ~/.ansible.cfg
+          echo "ssh_args = -o ServerAliveInterval=60" >> ~/.ansible.cfg
+          mkdir -p ~/.ssh
+          touch ~/.ssh/config
```
Comment (Neudrino) [inline .github/workflows/build_all.yml]: They seem to be superfluous and duplicated to above.

File: .github/workflows/build_all.yml [hunk 13]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+          chmod 600  ~/.ssh/config
+          echo "Host *" >> ~/.ssh/config
+          echo "    ServerAliveInterval 240" >> ~/.ssh/config
```
Comment (Neudrino) [inline .github/workflows/build_all.yml]: This should probably be in the [section above](https://github.com/magma/magma/pull/13889/files#diff-71a71e993dac5ebcaa1ad57402f04ee226b14121a8052e4f994706670fd65a1aR913) which claims to `Setup SSH Keys and known_hosts`

File: .github/workflows/build_all.yml [hunk 14]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+          cat  ~/.ssh/config
```
Comment (Neudrino) [inline .github/workflows/build_all.yml]: This seems like debug code and should be removed.

File: .github/workflows/build_all.yml [hunk 15]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+      - name: Run the ansible script to build on graviton
+        run: |
+          cd "${{ env.MAGMA_ROOT }}"/lte/gateway/deploy
+          ansible-playbook --extra-vars "aws_secret_key=${{ secrets.AWS_LF_SECRET_KEY }} aws_access_key=${{ secrets.AWS_LF_ACCESS_KEY }} magma_registry_username=${{ secrets.JFROG_USERNAME }} magma_registry_password=${{ secrets.JFROG_PASSWORD }} magma_tag=${TAG} magma_branch=${GITHUB_REF##*/}" arm_build.yml
```
Comment (Neudrino) [inline .github/workflows/build_all.yml]: A one-argument-per-line-strategy would make the code much more readable and reviewable here.

File: .github/workflows/build_all.yml [hunk 16]
```diff
@@ -893,6 +893,88 @@ jobs:
...
+      - name: get the right metadata for firebase
+        id: publish_artifacts
+        run: |
+            ARTIFACTS="{\"packages\":[\"$MAGMA_AGW_CONTAINER_REGISTRY/agw-gateway_c:${TAG}\", \"$MAGMA_AGW_CONTAINER_REGISTRY/agw-gateway_python:${TAG}\"],\"valid\":true}"
+            echo "::set-output name=artifacts::$(echo $ARTIFACTS)"
+      - name: Notify failure to Slack
+        if: failure() && github.event_name == 'push'
+        uses: rtCamp/action-slack-notify@12e36fc18b0689399306c2e0b3e0f2978b7f1ee7 # pin@v2.2.0
+        # yamllint enable
+        env:
```
Comment (Neudrino) [inline .github/workflows/build_all.yml:962]: `SLACK_CHANNEL` is missing here?

**Review by Neudrino — CHANGES_REQUESTED** [review]
- Semantic PR check fails and should be fixed.
- DCO check fails and should be fixed.
- yamllint should be made happy.
- The test plan should be more more elaborate:
   - It should contain a link to at least one successful run of the given pipeline.
   - It should point to the correctly uploaded images with appropriate tags.

File: lte/gateway/deploy/arm_build.yml (hunk 2 above)
Comment (tmdzk) [reply lte/gateway/deploy/arm_build.yml:62]: I don't know any clean way to do that in Ansible, I could just re-execute the ansible again and again while there is no ip but that's not clean either

File: lte/gateway/deploy/arm_build.yml (hunk 3 above)
Comment (tmdzk) [reply lte/gateway/deploy/arm_build.yml]: that's an easy and lazy  way to fetch the metadata of that instance 
will change for: 
```
- ec2_instance_info:
    instance_ids:
      - i-12345678
```

File: lte/gateway/deploy/arm_build.yml (hunk 4 above)
Comment (tmdzk) [reply lte/gateway/deploy/arm_build.yml:96]: https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#registering-variables

Register is just an ansible action that save the output of a task in a variable, I could have called it whatever I want. 
As I use a specific naming  (name: "`arm-build-{{ magma_branch }}-{{ magma_tag }}"`)it will always return the same instance, that's why I can hardcode the first element of the array

File: lte/gateway/deploy/arm_build.yml (hunk 5 above)
Comment (tmdzk) [reply lte/gateway/deploy/arm_build.yml:158]: Cause that version of compose is not available through apt

File: .github/workflows/build_all.yml (hunk 6 above)
Comment (tmdzk) [reply .github/workflows/build_all.yml:903]: sure will do in next commit

File: .github/workflows/build_all.yml (hunk 12 above)
Comment (tmdzk) [reply .github/workflows/build_all.yml]: will factorize

**Comment by tmdzk** [comment]
@jheidbrink why did you tag 1.8 ? how do we differentiate 1.8.0 from 1.8.1?

File: .github/workflows/build_all.yml (hunk 6 above)
Comment (tmdzk) [reply .github/workflows/build_all.yml:903]: All jobs are using v2 seems a bit odd to have one job using version 3, we should bump every job for consistency, will leave v2 for now but let's keep in mind that we can upgrade

File: .github/workflows/build_all.yml (hunk 16 above)
Comment (tmdzk) [reply .github/workflows/build_all.yml:962]: Every slack token is paired to a default channel so technically it's not missing, but will add it

**Comment by jheidbrink** [comment]
> @jheidbrink why did you tag 1.8 ? how do we differentiate 1.8.0 from 1.8.1?

I tagged 1.8 to be consistent with the images for amd64 architecture: https://artifactory.magmacore.org/ui/repos/tree/General/docker-test%2Fagw_gateway_c%2F1.8

I think 1.8  is also used for orc8r images.

For agwc it is set in https://github.com/magma/magma/blob/8dc6dea56b429c06fe595d4ceb63b15913338734/.github/workflows/build_all.yml#L454

I think tagging `1.8.0` and `1.8.1` should be done by promiting the images.

File: .github/workflows/build_all.yml (hunk 11 above)
Comment (jheidbrink) [reply .github/workflows/build_all.yml:937]: I agree, especially since with the current overwriting, the `$GITHUB_ENV` will contain something like
```
...
TAG=1.8.0
TAG=abcd1234
```

While it seems that the latter is used, it is not that obvious to everyone reading the code. I think the more explicit `if...else` makes it more clear.

File: .github/workflows/build_all.yml [hunk 17]
```diff
@@ -893,6 +893,93 @@ jobs:
...
+          echo "Host *" >> ~/.ssh/config
+          echo "    ServerAliveInterval 240" >> ~/.ssh/config
+      - name: Install Dependencies
+        run: |
+          pip install ansible awscli boto3
+          sudo apt-get update
+      - name: Get the right tagging
+        if: github.event_name == 'push'
+        run: |
+          echo TAG="${MAGMA_VERSION}" >> $GITHUB_ENV
```
Comment (jheidbrink) [inline .github/workflows/build_all.yml:934]: All the other jobs in this workflow use
```
          if [ "$GITHUB_REF" = "refs/heads/master" ]; then
            echo TAG="${GITHUB_SHA:0:8}" >> $GITHUB_ENV
          else
            GIT_BRANCH_VERSION=${GITHUB_BASE_REF:-${GITHUB_REF#refs/heads/}}
            echo TAG="${GIT_BRANCH_VERSION:1}" >> $GITHUB_ENV
          fi
```

I think we can just copy that. It can handle more release branches than v1.8, and it prevents accidentally tagging something with `1.8.0` which is not from `v1.8` branch.

File: .github/workflows/build_all.yml [hunk 18]
```diff
@@ -893,6 +893,93 @@ jobs:
...
+          if [ "$GITHUB_REF" = "refs/heads/master" ]; then
+            echo TAG="${GITHUB_SHA:0:8}" >> $GITHUB_ENV
+          fi
+      # - name: Ansible config setup
+      #   run: |
+      #     echo "[defaults]" >> ~/.ansible.cfg
+      #     echo "host_key_checking = False" >> ~/.ansible.cfg
+      #     echo "ssh_args = -o ServerAliveInterval=60" >> ~/.ansible.cfg
+      - name: Run the ansible script to build on graviton
```
Comment (jheidbrink) [inline .github/workflows/build_all.yml:943]: This will probably only work on pushes (due to unavailable secrets in PRs), so we should guard it with an `if: github.event_name == 'push'`. In the long run we should adapt the job to support building-but-not-pushing when run without the necessary secrets.

**Comment by voisey** [comment]
@tmdzk What's the status of this PR?

**Comment by tmdzk** [comment]
> @tmdzk What's the status of this PR?

I need to finish another PR then I'll get back to that one

**Comment by nstng** [comment]
Hi @tmdzk, how should we proceed with this PR (incl issue)? My understanding is that we have no AWS resources to run an arm pipeline. Are there alternative resources to run this? If not, imo, the current state should be documented and archived.

**Comment by tmdzk** [comment]
> Hi @tmdzk, how should we proceed with this PR (incl issue)? My understanding is that we have no AWS resources to run an arm pipeline. Are there alternative resources to run this? If not, imo, the current state should be documented and archived.

The only way I see would be to wait for github action to offer arm runners (maybe through the last MacOs versions)
I think we can document and drop that PR (unfortunately )
